x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/indexlifecycle/LifecycleExecutionState.java

```diff
@@ -0,0 +1,253 @@
+    private static final String PHASE_DEFINITION = "phase_definition";
+
+    private final String phase;
+    private final String action;
+    private final String step;
+    private final String failedStep;
+    private final String stepInfo;
+    private final String phaseDefinition;
+    private final long indexCreationDate;
+    private final long phaseTime;
+    private final long actionTime;
+    private final long stepTime;
+
+    private LifecycleExecutionState(String phase, String action, String step, String failedStep,
+                                    String stepInfo, String phaseDefinition, long indexCreationDate,
+                                    long phaseTime, long actionTime, long stepTime) {
+        this.phase = phase;
+        this.action = action;
+        this.step = step;
+        this.failedStep = failedStep;
+        this.stepInfo = stepInfo;
+        this.phaseDefinition = phaseDefinition;
+        this.indexCreationDate = indexCreationDate;
+        this.phaseTime = phaseTime;
+        this.actionTime = actionTime;
+        this.stepTime = stepTime;
+    }
+
+    /**
+     * Retrieves the execution state from an {@link IndexMetaData} based on the
+     * custom metadata.
+     * @param indexMetaData The metadata of the index to retrieve the execution
+     *                      state from.
+     * @return The execution state of that index.
+     */
+    public static LifecycleExecutionState fromIndexMetadata(IndexMetaData indexMetaData) {
+        Map<String, String> customData = indexMetaData.getCustomData(ILM_CUSTOM_METADATA_KEY);
+        customData = customData == null ? new HashMap<>() : customData;
+        return fromCustomMetadata(customData);
+    }
+
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    public static Builder builder(LifecycleExecutionState state) {
+        return new Builder()
+            .setPhase(state.phase)
+            .setAction(state.action)
+            .setStep(state.step)
+            .setFailedStep(state.failedStep)
+            .setStepInfo(state.stepInfo)
+            .setPhaseDefinition(state.phaseDefinition)
+            .setIndexCreationDate(state.indexCreationDate)
+            .setPhaseTime(state.phaseTime)
+            .setActionTime(state.actionTime)
+            .setStepTime(state.stepTime);
+    }
+
+    static LifecycleExecutionState fromCustomMetadata(Map<String, String> customData) {
+        return new Builder()
+            .setPhase(customData.getOrDefault(PHASE, ""))
+            .setAction(customData.getOrDefault(ACTION, ""))
+            .setStep(customData.getOrDefault(STEP, ""))
+            .setFailedStep(customData.getOrDefault(FAILED_STEP, ""))
+            .setStepInfo(customData.getOrDefault(STEP_INFO, ""))
+            .setPhaseDefinition(customData.getOrDefault(PHASE_DEFINITION, ""))
+            .setIndexCreationDate(Long.parseLong(customData.getOrDefault(INDEX_CREATION_DATE, "-1")))
```

Comment:
Could you add more descriptive parsing error messages in the event that these are incorrectly formatted as a long (ie, including the name)